### ISSUE TITLE
Duplicate approvals

### ISSUE THREAD

**Issue by JacobHayes** [hmarr/auto-approve-action#213]
After making a change to an approved path, every subsequent commit triggers a new approval. The logs say it is looking for existing reviews, but seems to not find the last one.

<img width="640" alt="Screenshot 2023-02-28 at 19 17 13" src="https://user-images.githubusercontent.com/2555532/222012764-e48a0ab4-061e-462d-9973-878b5aed2d64.png">

Logs from the second run:
```
Fetching user, pull request information, and existing reviews
Current user is github-actions[bot]
Commit SHA is AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
Pull request #123 has not been approved yet, creating approving review
Approved pull request #123
```
workflow config:
```
name: Auto approval
on:
  pull_request:
    branches:
      - main
    paths:
      - 'some-dir/**'
jobs:
  auto-approve:
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    steps:
      - uses: hmarr/auto-approve-action@v3
        with:
          review-message: "Automatic approval of ..."
```

---

Perhaps there's a missing permission that is preventing it from fetching existing reviews?

**Comment by hmarr** [comment]
I believe the default behaviour is to re-approve when new commits are pushed to the branch (`commit_id == prHead`):

https://github.com/hmarr/auto-approve-action/blob/aa8a692eccc605be7992f8a5cbe41893f72ecbc3/src/approve.ts#L39-L45

I assume that's so that re-reviews occur when they are required by branch protections:

<img width="785" alt="image" src="https://user-images.githubusercontent.com/110275/223824259-5346e50a-c60f-44ac-965c-7ddd1a8146f1.png">

However, the `state === "APPROVED"` check should cover that case, as when that box is checked, existing reviews transition to `DISMISSED` when new commits are pushed.

I'll try removing the check in a PR.

**Comment by hmarr** [comment]
Released in v3.2.0

**Comment by pfuhrmann** [comment]
I do not believe this works correctly now.

I have `DISMISSED` reviews, and the action is not re-approving PRs (it states that the PR was already approved by the user, which is true, but the review is dismissed and it seems to not detect that state).

Previously it worked correctly (albeit yes, it was duplicating sometimes, but I did not mind that).

**Comment by hmarr** [comment]
Thanks @pfuhrmann. It looks like it's possible to have old `APPROVED` reviews with newer `DISMISSED` reviews, and only the latest review for a given user counts towards a PR's review state, which is why this is happening. I've opened https://github.com/hmarr/auto-approve-action/pull/216 to fix that.

**Comment by hmarr** [comment]
@pfuhrmann if you could confirm that you're seeing the same thing (old `APPROVED` reviews with newer `DISMISSED` reviews for a given user), that'd be helpful.

You can see all the reviews by running `gh api repos/{LOGIN}/{REPO}/pulls/{PR-NUMBER}/reviews` if you have the [gh cli](https://github.com/cli/cli) installed.

**Comment by pfuhrmann** [comment]
Hi @hmarr, thanks for coming back to me so quickly!

This is what I see on the PR (redacted):
```json
[
  {
    "id": 1340392942,
    "node_id": "PRR_kwDOH65R-s5P5MXu",
    "user": {
      "login": "github-actions[bot]",
    },
    "body": "Auto approved (low risk update)! :+1: :rocket:",
    "state": "APPROVED",
    "submitted_at": "2023-03-14T22:41:54Z",
    "commit_id": "2e0af399a670ff3971253f69539274b1dedf13a1"
  },
  {
    "id": 1340392945,
    "node_id": "PRR_kwDOH65R-s5P5MXx",
    "user": {
      "login": "github-actions[bot]",
    },
    "body": "Auto approved (low risk update)! :+1: :rocket:",
    "state": "DISMISSED",
    "submitted_at": "2023-03-14T22:41:55Z",
    "commit_id": "2e0af399a670ff3971253f69539274b1dedf13a1"
  }
]

```

The full output from action is:
```bash
Fetching user, pull request information, and existing reviews
Current user is github-actions[bot]
Commit SHA is 2968fc1548df680b12c727068ad3f9075c56294f
Current user already approved pull request #77, nothing to do
```

I assume this is the state you wrote about previously, i.e. old `APPROVED` reviews with newer `DISMISSED` reviews. Seem's `DISMISSED` PR `submitted_at` is 1s later than `APPROVED`. Probably an edge case, and can happen if there are multiple actions running in parallel?

**Comment by hmarr** [comment]
Great, yeah that does look like the case I described. I've merged the PR and released v3.2.1 — hopefully that fixes the issue for you!

**Comment by pfuhrmann** [comment]
@hmarr I can confirm, all our automated PRs got merged so this seems resolved. Thanks!
